pkg/logql/syntax/ast.go

```diff
@@ -544,12 +544,12 @@ func newNestedLineFilterExpr(left *LineFilterExpr, right *LineFilterExpr) *LineF
 
 func (e *LineFilterExpr) Walk(f WalkFn) {
 	f(e)
-	if e.Left != nil {
-		e.Left.Walk(f)
-	}
-	if e.Or != nil {
-		e.Or.Walk(f)
-	}
+	// if e.Left != nil {
+	// 	e.Left.Walk(f)
+	// }
+	// if e.Or != nil {
+	// 	e.Or.Walk(f)
+	// }
```

Comment:
That is needed because otherwise each branch of the LineFilterExpr is visited adding them multiple times to the logical plan.

Reply:
I'm worried this is breaking the contract of ast.Walk for our specific use case; generally walking _should_ traverse children nodes. Can we find a way to avoid commenting this out?

Reply:
I agree. I haven't found another solution yet, though.

Reply:
How much effort would it take to update `WalkFn` so that it return a value (bool, error, nil, something) which indicates whether children should be walked or not? That would match the pattern used by the stdlib's [ast.Walk](https://pkg.go.dev/go/ast#Walk) and [fs.WalkDirFunc](https://pkg.go.dev/io/fs#WalkDirFunc).

Reply:
Other implementations of Walk often use a boolean return value to indicate whether the Walk should continue traversing it's children. But that would change any usages of the existing Walk

Reply:
https://github.com/grafana/loki/pull/16916/commits/35f2b144d16e538f7010ceef4bcdaa132f9d3e41

Reply:
Opened a separate PR for it https://github.com/grafana/loki/pull/16989